File: witchhunt/lobby.py
import html
import logging
import secrets

from collections import defaultdict

from witchhunt.configuregame import ConfigureGame
from witchhunt.message import Message
from witchhunt.player import Player


class Lobby:
    def __init__(self):
        self.name = None
        self.players = []
        self.accepting_users = True
        self.backend = ConfigureGame(self)
        self.seq_id = 0
        self.button_callbacks = {}
        self.components = defaultdict(set)

    def __repr__(self):
        return f"{self.name=} {self.accepting_users=} {self.players=}"

    def seq(self):
        self.seq_id += 1
        return self.seq_id

    def login(self, address, lobby=None, username=None, password=None):
        """
        A bit special, this method returns (lobby, messages) so the top level function
        can associate this address with a lobby.
        """
        username = html.escape(username)

        messages = []

        lobby_error = "" if lobby else "Required"
        username_error = "" if username else "Required"
        password_error = "" if password else "Required"
        is_admin = False

        def error():
            messages.append(
                Message(
                    address,
                    {
                        "login_messages": {
                            "lobby": lobby_error,
                            "username": username_error,
                            "password": password_error,
                        },
                    },
                )
            )
            return None, messages

        def finish(previous_login=None):
            # Notify the existing player that it is being disconnected
            if previous_login:
                messages.extend(previous_login.update_address(address))
            else:
                messages.extend(self.backend.player_join(username, is_admin))
                # TODO: add message to everyone that someone has joined the lobby
                # TODO: update admin panel
                # TODO: update game selection actions (admin and non-admin)
                self.players.append(Player(address, username, password, is_admin, self))
            messages.append(
                Message(
                    address,
                    {
                        "logged_in": {
                            "lobby": lobby,
                            "username": username,
                            "password": password,
                        },
                    },
                )
            )
            return self, messages

        if any((lobby_error, username_error, password_error)):
            return error()

        if len(username) > 12:
            username_error = "Too long"

        if len(lobby) > 10:
            lobby_error = "Too long"

        if lobby != html.escape(lobby):
            lobby_error = "No html characters"

        if any((lobby_error, username_error, password_error)):
            return error()

        if not self.players:
            # This is the first player. Any username/password is valid.
            self.name = lobby  # Needs to be set lazily because defaultdict is dumb
            is_admin = True
            return finish()

        previous_login = [p for p in self.players if p.username == username]
        if not previous_login:
            if self.accepting_users:
                # FIXME: notify the backend a player has joined
                return finish()
            else:
                lobby_error = "Game in progress, not accepting new users"
                return error()

        assert len(previous_login) == 1
        previous_login = previous_login[0]

        if previous_login.password != password:
            password_error = "Incorrect password or username already taken"
            return error()

        return finish(previous_login)

    def click_button(self, address, field, value, client_seq_id):
        # value is needed here because we don't want to blindly toggle the callback,
        # we should have the user say what they expect to happen
        ret = []
        try:
            callback = self.button_callbacks[field]
        except KeyError:
            logging.warning("Unknown button clicked: " + field)
            return ret

        ret.extend(callback(value, address, client_seq_id))
        return ret

    def create_button(self, component, callback):
        while True:
            private_id = secrets.token_hex(6)
            if private_id not in self.button_callbacks:
                break
        self.button_callbacks[private_id] = callback
        self.components[component].add(private_id)
        return private_id

    def destroy_component(self, component):
        for private_id in self.components[component]:
            del self.button_callbacks[private_id]
        del self.components[component]

File: backend/src/main/java/org/kevinstock/witchhunt/Lobby.java
package org.kevinstock.witchhunt;

import com.google.common.html.HtmlEscapers;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public class Lobby {
    private static final Logger logger = LoggerFactory.getLogger(Lobby.class);

    private static final String PLAYER_STATUS = "player_status";
    private static final String LEAVE_LOBBY = "leave_lobby";

    private final String name;
    private final Map<String, Player> usernameLookup = new HashMap<>();
    private final Map<Player, String> kickPlayer = new HashMap<>();
    private final Map<Player, String> promotePlayer = new HashMap<>();
    private final Map<String, Consumer<Boolean>> actions = new HashMap<>();

    private int day = 0;
    private String phaseIcon = "config";
    private ConfigureGame configuration = new ConfigureGame(this);
    private boolean mutablePlayerList = true;

    public Lobby(String name) {
        this.name = name;
    }

    /*
    Make sure that the values are good enough that if this lobby doesn't exist, we can create it and log this user
    in as the admin.

    All the requireNonNull  here are just to make intellij be quiet because it's analysis doesn't see we'd have already
    returned.

    The unstable api is guava's HtmlEscapers. If this ends up being the only use of guava, just implement it here.
    Or try using proguard: https://github.com/google/guava/wiki/UsingProGuardWithGuava
     */
    @SuppressWarnings("UnstableApiUsage")
    public static boolean basicLoginCheck(ClientConnection client, String lobby, String username, String password) {
        LoginError loginError = new LoginError();

        if (lobby == null || lobby.equals("")) {
            loginError.lobby = "Required";
        }

        if (username == null || username.equals("")) {
            loginError.username = "Required";
        }

        if (password == null || password.equals("")) {
            loginError.password = "Required";
        }

        if (loginError.hasErrors(client)) {
            return false;
        }

        if (!Objects.requireNonNull(lobby).equals(HtmlEscapers.htmlEscaper().escape(lobby))) {
            loginError.lobby = "No html characters";
        }

        if (!Objects.requireNonNull(username).equals(HtmlEscapers.htmlEscaper().escape(username))) {
            loginError.username = "No html characters";
        }

        if (loginError.hasErrors(client)) {
            return false;
        }

        if (lobby.length() > 12) {
            loginError.lobby = "Too long, max 12 characters";
        }

        if (username.length() > 12) {
            loginError.username = "Too long, max 12 characters";
        }

        if (Objects.requireNonNull(password).length() > 60) {
            loginError.password = "Too long, max 60 characters";
        }

        return !loginError.hasErrors(client);
    }

    // Ok yes, I'm just lazy. All the entry methods here are sync because I don't want to worry about race conditions.
    synchronized public void login(ClientConnection client, String username, String password) {
        if (usernameLookup.containsKey(username)) {
            Player existing = usernameLookup.get(username);
            if (existing.getPassword().equals(password)) {
                existing.updateClient(client);
                existing.send("logged_in", new LoggedInMessage(this.name, username, password));
                updatePlayerStatus();
            } else {
                LoginError.passwordError(client, "Username taken / incorrect password");
            }
        } else if (!mutablePlayerList) {
            LoginError.lobbyError(client, "Not accepting new players right now");
        } else {
            Player player = new Player(client, this, username, password);
            usernameLookup.put(username, player);
            configuration.addParticipant(player);
            kickPlayer.put(player, createAction(ignored -> removePlayer(player)));
            promotePlayer.put(player, createAction(ignored -> makeAdmin(player)));
            new Buttons(
                    this,
                    LEAVE_LOBBY,
                    List.of("Leave lobby"),
                    List.of(ignored -> removePlayer(player)),
                    List.of(player)
            );
            if (usernameLookup.size() == 1) {
                player.makeAdmin();
                configuration.addWriter(player);
            }
            player.send("logged_in", new LoggedInMessage(this.name, username, password));
            player.sendSecretMessage(String.format("Welcome to lobby %s. Your username is '%s' and your password is '%s'.", name, username, password), List.of("join", "username"));
            sendPublicMessage(username + " has joined the lobby.", List.of("join", username));
            updateAdminButtons();
            updatePlayerStatus();
        }
    }

    synchronized public void set(ClientConnection client, int clientSeqId, String key, boolean value) {
        client.getPlayer().updateLatestClientId(clientSeqId);

        Consumer<Boolean> callback = actions.get(key);
        if (callback == null) {
            logger.error("Unknown action key: {}", key);
        } else {
            callback.accept(value);
        }
    }

    public int getPlayerCount() {
        return usernameLookup.size();
    }

    public String getPhaseIcon() {
        return phaseIcon;
    }

    public void setPhaseIcon(String icon) {
        phaseIcon = icon;
    }

    public int getDay() {
        return day;
    }

    public void setDay(int day) {
        this.day = day;
    }

    public void lockPlayerList() {
        mutablePlayerList = false;
    }

    public void sendAdminsMessage(String message, List<String> tags) {
        List<String> allTags = new ArrayList<>(tags);
        allTags.add("admin");
        usernameLookup.values().stream().filter(Player::isAdmin).forEach(player -> player.sendSecretMessage(message, allTags));
    }

    public void sendPublicMessage(String message, List<String> tags) {
        usernameLookup.values().forEach(player -> player.sendPublicMessage(message, tags));
    }

    public List<Buttons.ButtonMessage> getAdminButtons() {
        List<Buttons.ButtonMessage> buttons = new ArrayList<>();

        // TODO: add pause and advance phase buttons as needed

        // TODO: switch to kill player buttons once in game
        kickPlayer.forEach((player, button) -> buttons.add(new Buttons.ButtonMessage("Kick " + player.getUsername(), button)));

        // TODO: make these a different color
        promotePlayer.forEach(
                (player, button) -> {
                    if (!player.isAdmin()) {
                        buttons.add(
                                new Buttons.ButtonMessage(
                                        "Promote " + player.getUsername(), button));
                    }
                });

        return buttons;
    }

    private void updateAdminButtons() {
        usernameLookup.values().forEach(Player::sendAdminButtons);
    }

    public void updatePlayerStatus() {
        List<PlayerStatusMessage> status = usernameLookup
                .values()
                .stream()
                .map(p ->
                        new PlayerStatusMessage(p.getUsername(), p.isConnected(), p.isAlive(), p.isAdmin()))
                .sorted(Comparator.comparing(PlayerStatusMessage::getUsername, String.CASE_INSENSITIVE_ORDER))
                .collect(Collectors.toList());
        usernameLookup.values().forEach(p -> p.send(PLAYER_STATUS, status));

        // TODO: Make sure this is called when a player dies
    }

    public void makeAdmin(Player player) {
        player.makeAdmin();
        configuration.addWriter(player);
        updatePlayerStatus();
        updateAdminButtons();
    }

    public void removePlayer(Player player) {
        if (!mutablePlayerList) {
            return;
        }

        kickPlayer.remove(player);
        promotePlayer.remove(player);
        configuration.removePlayer(player);

        for (String username : usernameLookup.keySet()) {
            if (usernameLookup.get(username).equals(player)) {
                usernameLookup.remove(username);
                break;
            }
        }

        sendPublicMessage(player.getUsername() + " has left the lobby.",
                List.of("join", player.getUsername()));

        // Promote someone to admin if there isn't one
        if (usernameLookup.values().stream().noneMatch(Player::isAdmin)) {
            usernameLookup.values().stream().findFirst().ifPresent(Player::makeAdmin);
        }

        player.disconnect();
        updateAdminButtons();
        // This doesn't need to update player status because the player's
        // connection will be closed, which triggers a status update
    }

    public void resetLobby() {
        actions.clear();
        day = 0;
        phaseIcon = "config";
        mutablePlayerList = true;

        configuration = new ConfigureGame(this);

        usernameLookup.values().forEach(player -> {
            configuration.addParticipant(player);
            if (player.isAdmin()) {
                configuration.addWriter(player);
            }
        });
    }

    public String createAction(Consumer<Boolean> callback) {
        String selector = UUID.randomUUID().toString();
        actions.put(selector, callback);
        return selector;
    }

    @SuppressWarnings({"FieldCanBeLocal", "UnusedDeclaration", "MismatchedQueryAndUpdateOfCollection"})
    private static class LoggedInMessage {
        private final String lobby;
        private final String username;
        private final String password;

        private LoggedInMessage(String lobby, String username, String password) {
            this.lobby = lobby;
            this.username = username;
            this.password = password;
        }
    }

    private static class LoginError {
        String lobby = "";
        String username = "";
        String password = "";

        public static void passwordError(ClientConnection client, String message) {
            LoginError error = new LoginError();
            error.password = message;
            error.hasErrors(client);
        }

        public static void lobbyError(ClientConnection client, String message) {
            LoginError error = new LoginError();
            error.lobby = message;
            error.hasErrors(client);
        }

        public boolean hasErrors(ClientConnection client) {
            boolean error = !(lobby.equals("") && username.equals("") && password.equals(""));
            if (error) {
                client.send("login_messages", this);
            }
            return error;
        }
    }

    @SuppressWarnings({"FieldCanBeLocal", "UnusedDeclaration", "MismatchedQueryAndUpdateOfCollection"})
    private static class PlayerStatusMessage {
        private final String username;
        private final boolean connected;
        private final boolean alive;
        private final boolean admin;

        private PlayerStatusMessage(String username, boolean connected, boolean alive, boolean admin) {
            this.username = username;
            this.connected = connected;
            this.alive = alive;
            this.admin = admin;
        }

        public String getUsername() {
            return username;
        }
    }
}
